homestead.root/usr/share/clearwater/bin/poll_cx.py

```diff
@@ -0,0 +1,287 @@
+#
+# The rationale for this function is that, if many diameter requests are failing
+# with server errors and no requests are succeeding, this _might_ be because
+# the Diameter stack is broken (rather than a genuine server or configuration
+# issue) perhaps because freeDiameter failed to handle an unsual network
+# failure and "hung" the Diameter connections.  In cases such as this, it is
+# likely that service can be restored by restarting homestead, and there is no
+# harm in trying this, as no requests are succeeding anyway.
+#
+# Note that the script does the stat lookup and calculation twice (and only
+# responds in the positive if both checks suggest that the stack might be
+# broken).  The reasoning here is that its not possible to check the "Previous
+# 5 minute" value for all statistics atomically and therefore the values of
+# some stats might be for a different 5 minute period.  If the script only
+# checked once, it would be vulnerable to this "tick rollover" problem and
+# might incorrectly identify the stack as needing a reboot, invalidly
+# impacting service.
+
+import sys
+import re
+import netsnmp
+import operator
+import logging
+
+# OID template for that segment of a Cx stat table covering previous 5 minute
+# counts
+table_segment_head = '.1.2.826.0.1.1578918.9.5.%d.1.4.3'
+
+#logging.basicConfig(level=logging.DEBUG)
+_log = logging.getLogger(__name__)
+
+# Get tolerance to server errors. We will only treat the number of server
+# errors at the Diameter interface as serious enough to warrant a restart
+# (provided there have been no successes over the same period) if the total
+# number of errors exceeds this tolerance.
+server_error_tolerance = int(sys.argv[1])
+_log.info("Server error tolerance = {}".format(server_error_tolerance))
+
+# Get a connection to the local SNMP agent
+ss = netsnmp.Session(DestHost='localhost', Version=2, Community='clearwater')
+
+# Compile the "find last element in an OID" regex
+find_last_element = re.compile('([^\.]*)$')
+
+# This routine takes tuples of response codes (successful responses,
+# client error responses and server error responses) and returns the number of
+# success and client error responses received from Diameter over the previous
+# 5 minute interval.
+#
+# Any responses seen for return codes not in the tuples passed are
+# considered server errors
+#
+# INPUT
+#   head_oid         - Pre-indexed OID of the segment of the table we are
+#                      searching. This will be the OID of the Cx response
+#                      table concatenated with:
+#                      -  ".1.4" (".1" says "entry" and ".4" says "count element")
+#                      -  ".3" ("Previous 5 minute values")
+#                      -  AppID.  This will be ".0" for "Base" and ".1" for _3GPP
+#   successes        - The response codes in this segment that correspond to
+#                      2XX "success" HTTP codes from a Homestead user's
+#                      perspective
```

Comment:
Understood.  I've changed as per our agreed strategy below (all experimental and 2XXX and 4XXX result codes mean success, anything else means potential connection failure)